examples/sat/schoening.py

```diff
@@ -1,85 +1,165 @@
 """This module provides the Schöning example for solving SAT instances."""
 from typing import Optional
 import numpy as np
-import itertools
+import attrs
 
-from qubrabench.algorithms.search import search
-from qubrabench.benchmark import oracle
+from qubrabench.algorithms.search import (
+    search_by_sampling_with_replacement,
+    SamplingDomain,
+)
+from qubrabench.benchmark import oracle, named_oracle
 
 from sat import SatInstance, Assignment
 
-__all__ = ["schoening_solve"]
+__all__ = ["schoening_solve", "schoening_bruteforce_steps"]
+
+
+@oracle
+def schoening_with_randomness(
+    inst: SatInstance, initial_assignment: np.ndarray, walk_steps: np.ndarray
+) -> Optional[Assignment]:
+    """
+    Run Schoening's algorithm with fixed random choices.
+
+    Args:
+        initial_assignment: starting assignment (bit string)
+        walk_steps: choices to take at each step of the random walk
+        inst: The 3-SAT Instance for which to find a satisfying assignment.
+
+    Returns:
+        Satisfying assignment if found, None otherwise.
+    """
+    assignment = np.copy(initial_assignment)
+
+    # FIXME: get_stats(inst.evaluate) won't work, as underlying matrix can't be hashed.
+    eval = (named_oracle("inst.evaluate"))(lambda k: inst.evaluate(k))
```

Comment:
Another note: `inst.evaluate` returns a `TypeError: unhashable type 'csr_matrix'. In order to still bench oracle calls, I wrapped it here.

Reply:
can you extract this error as a standalone test and add it to `tests/test_benchmark.py`? (in a separate PR). That way I can improve the error message thrown by the API.